src/blockdiagonal.jl

```diff
@@ -143,3 +143,20 @@ function _block_indices(B::BlockDiagonal, i::Integer, j::Integer)
     end
     return p, i, j
 end
+
+
+import Base: copyto!, copy!
+
+
+function copy!(dest::BlockDiagonal, src::BlockDiagonal)
+    copyto!(dest, src)
+    return dest
+end
+
+function copyto!(dest::BlockDiagonal, src::BlockDiagonal)
```

Comment:
Can you explain why we need both `copy!` and `copyto!` / the benefit of them?

Reply:
Actually `copyto!` is probably useless. `copy!`, similarly to `mul!`, makes a lot of unnecessary allocations if not specialized. With the same matrices as defined in my other comment, I get

```
@benchmark copy!(b3, b1)

BenchmarkTools.Trial: 
  memory estimate:  8.68 MiB
  allocs estimate:  86400
  --------------
  minimum time:     3.782 ms (0.00% GC)
  median time:      5.339 ms (0.00% GC)
  mean time:        5.961 ms (11.78% GC)
  maximum time:     20.478 ms (38.04% GC)
  --------------
  samples:          838
  evals/sample:     1
```

Reply:
With specialized code:
```
BenchmarkTools.Trial: 
  memory estimate:  288 bytes
  allocs estimate:  4
  --------------
  minimum time:     1.180 μs (0.00% GC)
  median time:      1.265 μs (0.00% GC)
  mean time:        1.390 μs (0.00% GC)
  maximum time:     17.230 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10
```